text/0003-placeholders-in-conditions.md

```diff
@@ -0,0 +1,145 @@
+);
+```
+The only valid uses of `?resource` in the policy scope are `resource == ?resource` and `resource in ?resource`, and this RFC is not proposing to change that. This RFC also disallows the use of `?principal` and `?resource` in the condition if they don't _also_ appear in the scope. 
+
+Both restrictions are in place because these two placeholder variables have special meaning for indexing policies in a policy store. In particular, looking _only_ at the linking information (i.e., to what `?principal` and/or `?resource` are bound) is sufficient to precisely select linked policies for particular requests. If the variables could only be bound in the condition or in other positions in the scope, and used in arbitrary contexts, it would lead to selecting superfluous policies. 
+
+Given these restrictions, the above policy would be impossible to write without support for placeholder variables beyond `?resource` and `?principal`. With their support, we can write: 
+```
+template (?lowerbound)
+permit(principal, action, resource)
+when { ?lowerbound in resource };
+```
+Here, the `template (?lowerbound)` part _introduces_ a fresh placeholder variable which can be used legally within the policy template that follows. Eliding the `template (?lowerbound)` prefix would be illegal, signaling an unbound variable. 
+
+Multiple variables can be introduced together, separated by commas, e.g.,
+```
+template (?user, ?resourcebound)
+permit(principal, action, resource)
+when { 
+     ?resourcebound in resource 
+  && principal in ?user.delegates
+};
+```
+
+Policy writers can choose to include `?principal` and/or `?resource` in the list of template variables if they like, but doing so is not required. These two variables retain their special status.
+
+If a variable is included in the list that does not appear in the actual template, we can issue a warning.
+
+All introduced placeholder variables presented so far are assumed to be entities (just like `?principal` and `?resource`). The particular _type_ of the entity is not given, for simplicity. Eliding the type poses a problem for validating templates on their own; we would have to wait to validate a template each time it is linked. However, link-time validation makes it straightforward to allow placeholder variables to have any type (`string`, `bool`, etc.), not just entity types. 
+
+Link-time validation has the benefit that it is lower overhead to write the templates, and potentially more flexible, since the particulars of the types don't need to be spelled out. New types added after the template is created that work with the template don't require changing the template.
+
+Link-time validation has three drawbacks, though.
+
+1. Fundamental errors in a template are not discovered until the template is first linked. E.g., `permit(principal in ?principal,action,resource) when { 1 < "hello" >}` will _always_ be invalid when linked, and it would be nice to know that as early as possible.
```

Comment:
Partial schema validation offers a way around this. We'd still do link-time typechecking, but would be able to report some errors earlier.

Reply:
Nice! If/when we accept this one, we can update the partial-schema RFC to mention this.

Reply:
+1 partial schema validation could make this a non-issue.

Reply:
We could analyze a fully linked policy template though, right?

Reply:
Yes, but we also want to analyze templates in their own right.  For example, you might want to know if there is a way to instantiate one of your templates so that this instantiation breaks a property you care about.